### PR TITLE
Add support for functions to resource types

### PR DESCRIPTION
**Pull request by anthony-c-martin**
Add the ability for `ResourceType` to directly expose functions. This builds on top of #377, and is a pre-requisite for Radius being able to adopt the mainline Bicep compiler + extension.

Instead of building on top of the `ResourceFunctionType` class (which must be registered in the index instead of as a property of a `ResourceType`), I opted to create a new type, which makes fewer assumptions about the implementation being Azure-based.

This change also introduces `FunctionType` as a first-class type, which could also potentially be used in other places (for example to implement provider-level functions).

### CODE REVIEW

File: src/bicep-types/test/integration/baselines/foo/index.json
Comment (jeskew) [inline src/bicep-types/test/integration/baselines/foo/index.json:1]: Should there be a `"functions"` key here? It's unclear how the function types would be discoverable.

Comment (anthony-c-martin) [reply src/bicep-types/test/integration/baselines/foo/index.json:1]: Pushed this change - it was just an issue with my test code. You should now see the "functions" property under the resource.

File: src/Bicep.Types.UnitTests/TypeSerializerTests.cs [hunk 1]
```diff
@@ -95,7 +96,10 @@ public void Different_types_can_be_serialized_and_deserialized()
             #pragma warning restore 618
             var objectType = factory.Create(() => new ObjectType("steven", new Dictionary<string, ObjectTypeProperty>(), null));
             var arrayType = factory.Create(() => new ArrayType(factory.GetReference(objectType)));
-            var resourceType = factory.Create(() => new ResourceType("gerrard", ScopeType.ResourceGroup|ScopeType.Tenant, ScopeType.Tenant, factory.GetReference(objectType), ResourceFlags.None));
+            var stringType = factory.Create(() => new StringType(true, 3, 10, "^foo"));
+            var functionParam = new FunctionTypeParameter("arg", factory.GetReference(stringType), null);
+            var resourceMethodType = factory.Create(() => new FunctionType("sayHi", factory.GetReference(stringType), new [] { functionParam }, null));
+            var resourceType = factory.Create(() => new ResourceType("gerrard", ScopeType.ResourceGroup|ScopeType.Tenant, ScopeType.Tenant, factory.GetReference(objectType), ResourceFlags.None, new [] { factory.GetReference(resourceMethodType) }));
```
Comment (rynowak) [inline src/Bicep.Types.UnitTests/TypeSerializerTests.cs:102]: What should the scope be for a non-Azure type?

Comment (anthony-c-martin) [reply src/Bicep.Types.UnitTests/TypeSerializerTests.cs:102]: I think `ScopeType.Unknown` may work, but I'm not 100%. I'm not making any changes to this logic here.

File: src/Bicep.Types/Concrete/FunctionType.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    public string Name { get; }
+
+    public ITypeReference Output { get; }
+
+    public IReadOnlyList<FunctionTypeParameter> Parameters { get; }
+
+    public string? Description { get; }
+}
+
+public class FunctionTypeParameter
```
Comment (rynowak) [inline src/Bicep.Types/Concrete/FunctionType.cs:24]: nit: is this a *type parameter of a function*? or the *type of a parameter of a function*?

eg: `Func<T>` vs `Func(T)`

File: src/Bicep.Types/Concrete/FunctionType.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace Azure.Bicep.Types.Concrete;
+
+public class FunctionType : TypeBase
+{
+    [JsonConstructor]
+    public FunctionType(string name, ITypeReference output, IReadOnlyList<FunctionTypeParameter> parameters, string? description)
```
Comment (rynowak) [inline src/Bicep.Types/Concrete/FunctionType.cs:12]: Is the name here the function name? or a name for the type?

File: src/Bicep.Types/Concrete/FunctionType.cs (hunk 2 above)
Comment (anthony-c-martin) [reply src/Bicep.Types/Concrete/FunctionType.cs:24]: The latter. Any suggestions for naming?

File: src/Bicep.Types/Concrete/ResourceType.cs [hunk 4]
```diff
@@ -15,8 +16,8 @@ public enum ResourceFlags
     public class ResourceType : TypeBase
     {
         [JsonConstructor]
-        public ResourceType(string name, ScopeType scopeType, ScopeType? readOnlyScopes, ITypeReference body, ResourceFlags flags)
-            => (Name, ScopeType, ReadOnlyScopes, Body, Flags) = (name, scopeType, readOnlyScopes, body, flags);
+        public ResourceType(string name, ScopeType scopeType, ScopeType? readOnlyScopes, ITypeReference body, ResourceFlags flags, IReadOnlyList<ITypeReference>? functions)
```
Comment (rynowak) [inline src/Bicep.Types/Concrete/ResourceType.cs:19]: Just checking my work, it looks like based on `IReadOnlyList<ITypeReference>? functions` The function name is part of the function's type signature, does that sound right? It sounds strange to me.

File: src/Bicep.Types/Concrete/FunctionType.cs (hunk 3 above)
Comment (anthony-c-martin) [reply src/Bicep.Types/Concrete/FunctionType.cs:12]: We don't really have a strong concept of naming types in Bicep - so the former. Good point about separating the two out though - it is confusing to have this be a property of the types.

File: src/Bicep.Types/Concrete/FunctionType.cs (hunk 2 above)
Comment (rynowak) [reply src/Bicep.Types/Concrete/FunctionType.cs:20]: `FunctionParameterType`

Comment (anthony-c-martin) [reply src/Bicep.Types/Concrete/FunctionType.cs]: Renamed to `FunctionParameter`. I avoided `FunctionParameterType` because it isn't a type.

**Review by rynowak — APPROVED** [review]

**Review by jeskew — APPROVED** [review]

**Pull request merged**